File: app2/src/main/java/ru/neverlands/anclient/model/KaznaSnapshot.java
package ru.neverlands.anclient.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Снимок последнего ответа казны.
 *
 * Хранится целиком в `info/<profile nick>/kazna/kazna.txt`, чтобы UI мог показать
 * кеш без сети, а последующие действия (`Собрать`) могли найти актуальные
 * action-link по UID.
 */
public final class KaznaSnapshot {
    public final long generatedAtMs;
    public final String sourceUrl;
    public final String currentWca;
    public final String currentCategoryTitle;
    public final List<KaznaCategory> categories;
    public final List<KaznaItem> items;

    public KaznaSnapshot(
            long generatedAtMs,
            String sourceUrl,
            String currentWca,
            String currentCategoryTitle,
            List<KaznaCategory> categories,
            List<KaznaItem> items) {
        this.generatedAtMs = generatedAtMs;
        this.sourceUrl = sourceUrl == null ? "" : sourceUrl;
        this.currentWca = currentWca == null ? "" : currentWca;
        this.currentCategoryTitle = currentCategoryTitle == null ? "" : currentCategoryTitle;
        this.categories = categories == null ? new ArrayList<>() : new ArrayList<>(categories);
        this.items = items == null ? new ArrayList<>() : new ArrayList<>(items);
    }

    public List<KaznaCategory> copyCategories() {
        return new ArrayList<>(categories);
    }

    public List<KaznaItem> copyItems() {
        return new ArrayList<>(items);
    }

    public List<KaznaItem> readonlyItems() {
        return Collections.unmodifiableList(items);
    }

    public KaznaItem findItemByUid(String uid) {
        if (uid == null || uid.trim().isEmpty()) {
            return null;
        }
        String normalized = uid.trim();
        for (KaznaItem item : items) {
            if (normalized.equals(item.uid)) {
                return item;
            }
        }
        return null;
    }
}

File: ANClient/Kazna/KaznaSnapshot.cs
using System;
using System.Collections.Generic;

namespace ANClient.Kazna
{
    internal sealed class KaznaSnapshot
    {
        internal DateTime GeneratedAt;
        internal string SourceUrl;
        internal string CurrentWca;
        internal string CurrentCategoryTitle;
        internal readonly List<KaznaCategory> Categories = new List<KaznaCategory>();
        internal readonly List<KaznaItem> Items = new List<KaznaItem>();

        internal KaznaItem FindItemByUid(string uid)
        {
            if (string.IsNullOrEmpty(uid))
                return null;

            for (var i = 0; i < Items.Count; i++)
            {
                if (string.Equals(Items[i].Uid, uid.Trim(), StringComparison.OrdinalIgnoreCase))
                    return Items[i];
            }

            return null;
        }
    }
}
